Feat/readme (#24)
EDMdesigner/plugin-interface: README.md

- **data:** you can pass static data to your plugin.
- **settings:** you can pass down custom settings that modify the look and operation of the plugin.
- **hooks:** you can pass down functions that the plugin will call at certain actions or events.

**Parameters in the second object**

The `initFullscreenPlugin` function creates and iframe based on the `src` provided, and appends it to the `parentElem`. The second parameter object contains information for the library to create the iframe and append it to your application DOM. 

- **id:** is the id which will represent the iframe.
- **src:** this is the iframe source as a string.
- **parentElem:** this is a query selector or HTMLElement that you want the plugin iframe to be inserted into. Default is `document.body`
- **beforeInit:** this function will run after the iframe is created and the container and iframe both can be reached inside of this.

	```js
	beforeInit({ container, iframe }) {
		// your code here
	}
	```

- **timeout:** this is a number in milliseconds. This defines how long should the init function wait for an answer from the providePlugin before throwing an error.

### Interface

> In the provided interface you will reach the following props:

Reviewer: Talán kicsit érthetőbb.

```
In the returned object you will get the following properties
```

```js
{
	_container: HTMLElement,
	_src: String,
	methods: Object,
	showSplashScreen: Function,
	hideSplashScreen: Function,
	show: Function,
	hide: Function,
	destroy: Function,
}
```
- **_container:** is and HTML element containing the plugin iframe
- **_src:** is the source of the plugin iframe

> - **methods:** through the methods object you can reach the plugins declared methods

Reviewer: ```suggestion
- **methods:** through the methods object you can reach the plugin's declared methods
```

- **destroy:** this function removes the iframe from the container
#### Splashscreen
If the `settings` param provided in the initialization object contains a `splashScreenUrl`, the plugin will have a separate iframe appended to the container, which you can show and hide with the provided `showSplashScreen` and `hideSplashScreen` functions.
#### Show / Hide
You can show and hide your created plugins with the provided `show` and `hide`. While using these functions, the plugin won`t be destroyed, it will keep its state while hidden.

> **Animations**
> The `show` can be called with the following parameters:
> ```js
> show({ x = "-100vw", y = "0px", opacity = 0.5, scale = 1, time = 500 })
> ```

Reviewer: ```suggestion
**Animations**

The `show` can be called with the following parameters:

```js
show({ x = "-100vw", y = "0px", opacity = 0.5, scale = 1, time = 500 })
```
```

Reviewer: kell egy üres sor az animations után hogy ne fusson össze a szöveg

> The `show` function provides an easy way to customize your show animation. With the provided parameters, you can set the default hidden state, described by coordinates, opacity, and scale of the plugin, along with the time of the animation. When the function is ran, the plugin will move to fullscreen view from that hidden position. The animation uses the `translate3d` css function. Likewise, the `hide` function moves the plugin back to its set hidden state.

Reviewer: ```suggestion
The `show` function provides an easy way to customize your show animation. With the provided parameters, you can set the default hidden state, described by coordinates, opacity, and scale of the plugin, along with the time of the animation. When the function is called, the plugin will move to a fullscreen view from that hidden position. The animation uses the `translate3d` css function. Likewise, the `hide` function moves the plugin back to its set hidden state.
```

- **timeout:** this is a number in milliseconds. This defines how long should the init function wait for an answer from the providePlugin before throwing an error.

### Interface

> In the provided interface you will reach the following props:

Reviewer: Talán kicsit érthetőbb.

```
In the returned object you will get the following properties
```

Reviewer: ez még mindig valid

```js

>  {
> 	_container: HTMLElement,
> 	methods: Object,
> 	destroy: Function,
> }

Reviewer: ```suggestion
{
	_container: HTMLElement,
	methods: Object,
	destroy: Function,
}
```

```
- **_container:** is and HTML element containing the plugin iframe

> - **methods:** through the methods object you can reach the plugins declared methods

Reviewer: ```suggestion
- **methods:** through the methods object you can reach the plugin's declared methods
```

});
```

- **hooks:** This is an array of hook names that the plugin accepts and uses
- **methods:** These are functions can be called from outside and are used to interact directly with the plugin from the outside

> - **validator:** Is a function that will run when the provided `data`, `settings` and `hooks` arrive from the from the parent side

Reviewer: ```suggestion
- **validator:** Is a function that will run when the provided `data`, `settings` and `hooks` arrive from the parent side
```

### Interface

> The providePlugin function should resolve to an object containing these fields:

Reviewer: Talán kicsit érthetőbb.

```
In the returned object you will get the following properties
```

```js
{
	data: Object,
	settings: Object,
	hooks: Object,
	terminate: Function,
}
```
- **data:** The data that was sent at the init stage
- **settings:** The settings that were sent at the init stage
- **hooks:** Hooks that were sent at the init stage and were filtered with the list of hooks that are accepted by the plugin
- **terminate:** A function designed to terminate the communication between the window objects.

## Example
### Parent side
```html
<!-- index.html -->
<html>
<body>
	<div id="pluginWrapper"></div>
</body>

> <script src="main.js"></script>

Reviewer: ezt az example-t ki is próbáltad?
mert elvileg itt ez kell hogy rendesen menjen:
```suggestion
<script src="main.js" type="module"></script>
```

</html>
```
```js
// main.js

> import { initFullscreenPlugin } from "@chamaileon-sdk/plugin-interface";
> const myPlugin = await initFullscreenPlugin({
> 	data: {},
> 	settings: {
> 		buttons: {
> 			header: [],
> 		},
> 	},
> 	hooks: {
> 		close: () => {
> 			myPlugin.hide();
> 		},
> 		onHeaderButtonClicked: ({ buttonId }) => {
> 			console.info("Button clicked: " + buttonId);
> 		},
> 	},
> }, {
> 	id: "myPlugin",
> 	src: "http://route/to/my/plugin/html",
> 	parentElem: "#pluginWrapper",
> 	beforeInit: () => {},
> 	timeout: 15000,
> });

Reviewer: ```suggestion
import { initFullscreenPlugin } from "@chamaileon-sdk/plugin-interface";
const myPlugin = await initFullscreenPlugin(
{
	data: {},
	settings: {
		buttons: {
			header: [],
		},
	},
	hooks: {
		close: () => {
			myPlugin.hide();
		},
		onHeaderButtonClicked: ({ buttonId }) => {
			console.info("Button clicked: " + buttonId);
		},
	},
},
{
	id: "myPlugin",
	src: "http://route/to/my/plugin/html",
	parentElem: "#pluginWrapper",
	beforeInit: () => {},
	timeout: 15000,
});
```

```
### Plugin side
```html
<html>
<body>
	I am a plugin
</body>

> <script src="plugin.js"></script>

Reviewer: ugyanaz a kérdés mint a másik script tag-nál

</html>
```
```js
// plugin.js
import { providePlugin } from "@chamaileon-sdk/plugin-interface";

> export function updateData(data = {}) {
> 	console.log(data);
> }

Reviewer: ```suggestion
function updateData(data = {}) {
	console.log(data);
}
```

> export function updateSettings(settings = {}) {
> 	console.log(settings);
> }

Reviewer: ```suggestion
function updateSettings(settings = {}) {
	console.log(settings);
}
```

});
```

- **hooks:** This is an array of hook names that the plugin accepts and uses
- **methods:** These are functions can be called from outside and are used to interact directly with the plugin from the outside
- **validator:** Is a function that will run when the provided `data`, `settings` and `hooks` arrive from the parent side
### Interface
The providePlugin function should resolve to an object containing these fields:
```js
{
	data: Object,
	settings: Object,
	hooks: Object,
	terminate: Function,
}
```
- **data:** The data that was sent at the init stage
- **settings:** The settings that were sent at the init stage
- **hooks:** Hooks that were sent at the init stage and were filtered with the list of hooks that are accepted by the plugin
- **terminate:** A function designed to terminate the communication between the window objects.

### Example

> You can run the examples, with static server
>  static-server -c "*" -zp 10101

Reviewer: ```suggestion
You can run the examples, with static server

```js
static-server -c "*" -zp 8080
```

```
